### PR TITLE
AN-970-add graph traversal source

### PR DESCRIPTION
**Pull request by xiazcy**
Added the base functions for graph traversal source, graph traversal, and traversal components, with stubbed functions where other components (such as bytecode) are needed.  Testing will be done during integration with these other components. 

Also made some nit fixes to previous testing files. 

@lyndonb-bq @simonz-bq @alexey-temnikov @L0Lmaker

### CODE REVIEW

File: gremlin-go/driver/connection_test.go [hunk 1]
```diff
@@ -29,22 +50,22 @@ func TestConnection(t *testing.T) {
 			assert.NotNil(t, resultSet)
 			result := resultSet.one()
 			assert.NotNil(t, result)
-			assert.Equal(t, result.AsString(), "[0]")
+			assert.Equal(t, "[0]", result.AsString())
```
Comment (alexey-temnikov) [inline gremlin-go/driver/connection_test.go]: Shall we improve best practices around testing by adding messages to asserts? If tests get broken, it is easier to find the root cause from the test report if additional context is provided. Example from: https://pkg.go.dev/github.com/stretchr/testify/assert#Equal
`assert.Equal(t, a, b, "The two words should be the same.")`

File: gremlin-go/driver/connection_test.go [hunk 2]
```diff
@@ -8,19 +27,21 @@ import (
...
 	t.Run("Test connect", func(t *testing.T) {
 		if runIntegration {
-			connection := connection{"localhost", 8181, Gorilla, newLogHandler(&defaultLogger{}, Info, language.English), nil, nil, nil}
+			connection := connection{testHost, testPort, Gorilla, newLogHandler(&defaultLogger{}, Info, language.English), nil, nil, nil}
 			err := connection.connect()
 			assert.Nil(t, err)
 		}
 	})
 
 	t.Run("Test write", func(t *testing.T) {
```
Comment (alexey-temnikov) [inline gremlin-go/driver/connection_test.go]: The test name is "write", but we do read operation (count). Suggesting to improve test naming convention to better reflect what test is doing.

**Comment by alexey-temnikov** [comment]
Minor nit: Can we align test naming convention along the way. These files are using the word 'Test', e.g. `"Test connect"`, while other unit tests don't. Perhaps we can align on the same style.

File: gremlin-go/driver/graphTraversal.go [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+func NewGraphTraversal(graph *Graph, traversalStrategies *TraversalStrategies, bytecode bytecode) *GraphTraversal {
+	return &GraphTraversal{t: &Traversal{
+		graph:               graph,
+		traversalStrategies: traversalStrategies,
+		bytecode:            bytecode,
+	}}
+}
+
+// Clone make a copy of a traversal that is reset for iteration.
+func (g *GraphTraversal) Clone() *GraphTraversal {
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphTraversal.go:35]: Not sure if I would consider this a clone. Clone to me would require deep copy, this is just a different pointer with the same underlying memory.

File: gremlin-go/driver/protocol_test.go [hunk 4]
```diff
@@ -39,6 +39,6 @@ func Test(t *testing.T) {
 		request := makeStringRequest("1+1")
 		requestID, err := protocol.write(&request, nil)
 		assert.NotNil(t, err)
-		assert.Equal(t, requestID, "")
```
Comment (lyndonbauto) [inline gremlin-go/driver/protocol_test.go:42]: 😆

File: gremlin-go/driver/traversal.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+*/
+
+package gremlingo
+
+type Traverser struct {
```
Comment (lyndonbauto) [inline gremlin-go/driver/traversal.go]: Lets not put this in yet - as a user I might be confused when I find a public facing Traverse, but it isn't implemented.

File: gremlin-go/driver/graphTraversalSource.go [hunk 6]
```diff
@@ -19,5 +19,138 @@ under the License.
...
+type Graph struct {
+}
+
+type TraversalStrategies struct {
+}
+
+type GraphTraversalSource struct {
+	graph               *Graph
+	traversalStrategies *TraversalStrategies
+	bytecode            bytecode
```
Comment (simonz-bq) [inline gremlin-go/driver/graphTraversalSource.go]: Should this not be a pointer?

File: gremlin-go/driver/bytecode.go [hunk 7]
```diff
@@ -22,4 +22,5 @@ package gremlingo
 type bytecode interface {
 	addSource(sourceName string, args ...interface{})
 	addStep(stepName string, args ...interface{})
+	toString() string
```
Comment (simonz-bq) [inline gremlin-go/driver/bytecode.go]: Didn't I remove this in my PR?

This probably will change when you rebase.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/bytecode.go (hunk 7 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go]: Yes, that was still before the rebase.

File: gremlin-go/driver/graphTraversalSource.go [hunk 8]
```diff
@@ -19,5 +19,132 @@ under the License.
...
+	return source
+}
+
+// With provides a configuration to a traversal in the form of a key value pair
+// TODO: Add to this when implementing traversal strategies
+func (gts *GraphTraversalSource) With(key interface{}, value interface{}) *GraphTraversalSource {
+	source := gts.clone()
+	source.bytecode.addSource("withStrategies", key, value)
+	return source
+}
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphTraversalSource.go]: I suggest removing this until we have a proper implementation for our traversal strategies.
```suggestion
//func (gts *GraphTraversalSource) With(key interface{}, value interface{}) *GraphTraversalSource {
//	source := gts.clone()
//	source.bytecode.addSource("withStrategies", key, value)
//	return source
//}
```

**Review by lyndonbauto — APPROVED** [review]

**Comment by simonz-bq** [comment]
Please uncomment the commented codeblock in `bytecode.go`

**Pull request merged**